compiler/threading/hash-table.h

```diff
@@ -13,35 +13,35 @@ template<class T, int N = 1000000>
 class TSHashTable {
 public:
   struct HTNode : Lockable {
-    unsigned long long hash;
+    std::atomic<unsigned long long> hash;
     T data;
 
-    HTNode() :
-      hash(0),
-      data() {
-    }
+    HTNode()
+      : hash(0)
+      , data() {}
   };
 
 private:
   HTNode *nodes;
-  int used_size;
+  std::atomic<int> used_size;
+
 public:
-  TSHashTable() :
-    nodes(new HTNode[N]),
-    used_size(0) {
-  }
+  TSHashTable()
+    : nodes(new HTNode[N])
+    , used_size(0) {}
 
   HTNode *at(unsigned long long hash) {
     int i = (unsigned)hash % (unsigned)N;
     while (true) {
-      while (nodes[i].hash != 0 && nodes[i].hash != hash) {
+      while (nodes[i].hash.load(std::memory_order_acquire) != 0 && nodes[i].hash.load(std::memory_order_relaxed) != hash) {
```

Comment:
because it won't be reordered between `nodes[i].hash.load(std::memory_order_acquire)` before and `nodes[i].hash.load(std::memory_order_acquire)` after, the only barrier needed here is load-load